MagazineLayout/LayoutCore/SectionModel.swift

```diff
@@ -371,6 +373,29 @@ struct SectionModel {
   {
     indexOfFirstInvalidatedRow = min(proposedIndex, indexOfFirstInvalidatedRow ?? proposedIndex)
   }
+  
+  private mutating func applyRowOffsetsIfNecessary() {
+    guard let rowOffsetTracker = rowOffsetTracker else { return }
+
+    // Accessing this array using an unsafe, untyped (raw) pointer avoids expensive copy-on-writes
+    // and Swift retain / release calls.
+    let itemModelsPointer = UnsafeMutableRawPointer(mutating: &itemModels)
+    let directlyMutableItemModels = itemModelsPointer.assumingMemoryBound(to: ItemModel.self)
+
+    for rowIndex in 0..<numberOfRows {
+      let rowOffset = rowOffsetTracker.offsetForRow(at: rowIndex)
+      switch rowIndex {
+      case indexOfHeaderRow(): headerModel?.originInSection.y += rowOffset
+      case indexOfFooterRow(): footerModel?.originInSection.y += rowOffset
+      default:
+        for itemIndex in itemIndicesForRowIndices[rowIndex] ?? [] {
+          directlyMutableItemModels[itemIndex].originInSection.y += rowOffset
+        }
+      }
+    }
+
+    self.rowOffsetTracker = nil
+  }
```

Comment:
Here's the fix for the main bug from the original PR - the basic idea is that if we ever need to invalidate our row offsets (the number of rows changes due to some batch update, for example), then we need to apply the row offsets we know about since we're about to start fresh with a new row offset tracker.

In the future, we can make some optimizations to minimize how often this "apply" step needs to happen. For example, if a full width item (a row) is inserted, then we can just add the height of it to our row offsets. A similar optimization could be done if an entire row of item(s) is removed. These optimizations will require the `RowOffsetTracker` to support updating the segment tree (which stores the offsets) more efficiently.

For now though, this is still a huge improvement over the existing implementation. Scroll performance when self-sizing is still super fast (this apply code path is never even hit), and we only need to do this expensive "apply" step when doing batch updates.